synapse/logging/opentracing.py

```diff
@@ -914,36 +916,49 @@ async def _wrapper(
                 return await func(*args, **kwargs)  # type: ignore[misc]
 
     else:
-        # The other case here handles both sync functions and those
-        # decorated with inlineDeferred.
+        # The other case here handles sync functions including those decorated with
+        # `@defer.inlineCallbacks` or return a `Deferred`, and those that return
+        # `Awaitable`
         @wraps(func)
-        def _wrapper(*args: P.args, **kwargs: P.kwargs) -> R:
+        def _wrapper(*args: P.args, **kwargs: P.kwargs) -> Any:
             scope = wrapping_logic(func, *args, **kwargs)
             scope.__enter__()
 
             try:
                 result = func(*args, **kwargs)
+
                 if isinstance(result, defer.Deferred):
 
                     def call_back(result: R) -> R:
                         scope.__exit__(None, None, None)
                         return result
 
                     def err_back(result: R) -> R:
+                        # XXX: Feels like we could put the error details into the
+                        # `scope.__exit__(...)`
```

Comment:
For consistency with the sync and async coroutine paths, I think we _should_ do this, otherwise it's decorator behavior we can't rely upon.
```suggestion
                        # TODO: Pass the error details into `scope.__exit__(...)` for
                        #       consistency with the other paths.
```